### ISSUE TITLE
Ignore non-comments

### ISSUE THREAD

**Issue by dnut** [abrookins/WrapToColumn#29]
Any adjacent lines of code are grouped together with the comment paragraph. This should only work for comments.

Originally I made this ticket to ignore the shebang line, but then I realized it groups everything, comment or not.

**Comment by cjbarth** [comment]
I agree, when I go to format the paragraph of a docstring I go from:

```python
class MyClass:
    """
    This is a very long string that serves to document this function and is otherwise totally pointless.
    """"
```

to this:

```python
class MyClass: """ This is a very long string that serves to document this function and is otherwise
totally pointless. """"
```

When I'm expecting this:

```python
class MyClass:
    """
    This is a very long string that serves to document this function and is otherwise totally
    pointless.
    """"
```

I understand that the tool needs to include adjacent lines in order to reformat paragraphs that are broken up at weird line lengths to put them back to normal, however, it seems a special case should be made for docstrings being surrounded by `"""`.

**Comment by cjbarth** [comment]
It seems this could be trivially fixed by modifying the code [here](https://github.com/abrookins/WrapToColumn/blob/2d8157d0c13ae122f41fe176c8b311741225d1e5/src/com/andrewbrookins/idea/wrap/WrapParagraphAction.kt#L61). I'm not an expert in Java, but I'd be willing to take a stab at this if a PR would be welcome. @abrookins Do you have any thoughts on this?

@dnut How would you define an adjacent code line?

**Comment by dnut** [comment]
> @dnut How would you define an adjacent code line?

Adjacent:
```python
# comment
code()
```
Not adjacent:
```python
# comment

code()
```

I think the problem I was having was that the first example would get reformatted to something like:
```python
# comment code()
```
But now, I'm not having that problem. I think the behavior was changed such that it only reformats exactly the line that your cursor is on, or whatever lines are highlighted. Actually, I can't even reproduce your docstring example unless I highlight the entire thing.

**Comment by cjbarth** [comment]
Interesting. I find that if I use the "Wrap Line to Column" I don't have the problem, but if I use the "Wrap Paragraph to Column" I have the problem. Which are you using?

Also, I understand what adjacent code is as a human, but I'm wondering if you had any ideas to determine what adjacent code is using code.

**Comment by dnut** [comment]
Ah right. I have been using the ctrl-alt-shift-w which is the default keybind for  "wrap line to column."

Intellij is already smart enough to identify whether a specific line is a comment or not. My hope was that it exposes this to plugins, so the plugin could just check whether the adjacent line is a comment with something like an `isComment()` method.

If nothing like that exists, that complicates things. For single line comments it would need to check that the first non-whitespace characters designate a comment. For block comments it would need to parse the entire file to figure out exactly which lines are part of a block. In either case you would need to check that it's not part of a string or anything like that. In python there is a unique requirement that triple quoted strings are only considered comments in specific places. I'm sure there are other considerations that I haven't thought of. So it's not trivial.

**Comment by cjbarth** [comment]
Do you know where I might be able to find out what APIs Intellij exposes to plug-in developers? I've never tried this before, but I'd really love to try to get a PR together to fix this. It is actually quite sad that Intellij doesn't already do this when it wraps the rest of Python code.

**Comment by dnut** [comment]
I haven't looked into this before but a quick search turned this up: https://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/psi.html

You may want to look at the PSI source code. It seems well documented. https://upsource.jetbrains.com/idea-ce/structure/idea-ce-e97504227f5f68c58cd623c8f317a134b6d440b5/platform/core-api/src/com/intellij/psi

**Comment by abrookins** [comment]
Hey guys!

I haven't had time to go back and do much with this plugin since I wrote
it. The code is probably not super clear either -- the current incarnation
was something I wrote in a mad dash one Christmas while learning Kotlin,
which I never used again.

That said, the current problems the plugin has -- which you rightly called
out -- come from its rather simple regex-driven approach. It tries to infer
things from line prefixes without really knowing anything semantically
about the text it operates on; it just knows that "comments in some
languages start with these common line prefixes."

Intellij isn't like other editors that publish actual documented plugin
APIs and have instructions on how to use them. Like, you know, any other
text editor you can think of. Instead, in my experience, you have to read
the Community version of the editor's source to find out how anything
works, cross-reference that with published open-source plugins, and compare
those facts to the skimpy details published here and there and frequently
out of date in the official docs. It's also rather difficult to even set up
your editor to develop a plugin. Not for the faint of heart!

(Perhaps they've made it easier recently, but I highly doubt it.)

Anyway, I've always assumed there must be some way to ask the editor the
semantic meaning of text at a certain point in a document, but I haven't
had a chance to dig into that for ... I guess years now?

Definitely interested though, and would approve/advise on any PRs.

-A

**Comment by abrookins** [comment]
I'm back in the codebase after fixing a different issue, so I'll see if I can fix this problem this week.

**Comment by abrookins** [comment]
Should be fixed in 1.7.0, now published to the Plugins Marketplace.

It's not at the level that you can select comments and code and the plugin ignores the code, but we can at least stop looking for text to wrap with the Wrap Paragraph action if we see text that looks like it's probably code.

**Comment by cjbarth** [comment]
I may not be able to test this for a bit due to different work that keeps me out of IntelliJ products recently, but thank you for keeping this in mind and fixing it!
